### DISCUSSION TITLE
Does prec work with seq whose arity is greater than 3?

### DISCUSSION THREAD

**Discussion by panlina** [tree-sitter/tree-sitter#791] [Q&A] [answered]
My question is from my previous question: https://github.com/tree-sitter/tree-sitter/discussions/781.

Basically I defined comma expression like `x=1,x`, and equality comparison expression like `x=x`, and I want comma expression have lower precedence, thus parsing `x=1,x=x` as `(x=1),(x=x)`, instead of `(x=1,x)=x`:

```
module.exports = grammar({
	name: 'a',
	word: $ => $.identifier,
	rules: {
		expression: $ => choice(
			$.number,
			$.identifier,
			$.expression_binary,
			$.expression_comma
		),
		expression_comma: $ => prec.right(-1, seq(
			field('name', $.identifier),
			'=',
			field('value', $.expression),
			',',
			field('body', $.expression)
		)),
		expression_binary: $ => prec.left(0, seq(
			field('left', $.expression),
			field('operator', '='),
			field('right', $.expression)
		)),
		identifier: $ => /[_a-zA-Z][_a-zA-Z0-9]*/,
		number: $ => /\d+/,
	},
	conflicts: $ => [
		[$.expression, $.expression_comma],
		[$.expression_comma, $.expression_binary],
	]
});
```

but it gives following error when parsing `x=1,x=x`:

```
(expression [0, 0] - [0, 7]
  (expression_binary [0, 0] - [0, 7]
    left: (expression [0, 0] - [0, 3]
      (expression_binary [0, 0] - [0, 3]
        left: (expression [0, 0] - [0, 1]
          (identifier [0, 0] - [0, 1]))
        right: (expression [0, 2] - [0, 3]
          (number [0, 2] - [0, 3]))))
    (ERROR [0, 3] - [0, 5]
      (identifier [0, 4] - [0, 5]))
    right: (expression [0, 6] - [0, 7]
      (identifier [0, 6] - [0, 7]))))
example-file    0 ms    (ERROR [0, 3] - [0, 5])
```

Finally my solution was separating `expression_comma` into 2 rules, as described in [my comment](https://github.com/tree-sitter/tree-sitter/discussions#discussioncomment-119319), but I still don't know why I must do so.

My question is: does `prec` work with `seq` whose arity is greater than 3, or to put it in another way, does it work with a structure other than "operand-operator-operand"? If yes, What's wrong in my case?

Thanks.

**Comment by maxbrunsfeld** [comment, marked as answer]
Yes, it does the same thing regardless of the arity of the sequence.

In general, precedence tells Tree-sitter to *always* prefer to try parsing one rule over another. When using it, you often need to clearly understand *when* Tree-sitter would need to decide between the different conflicting options. The decision may be required earlier than you realize, causing the parser to commit too early to one alternative or another.

One tool that can be helpful in debugging this is running `tree-sitter parse -D`. It requires that `graphviz` be installed. It renders step-by-step graphs of what the parser is doing. By carefully following the sequence of actions, you may be able to see at what step it is making an unexpected decision, and then determine how precedence caused that decision.

**Reply by panlina** [reply]
I'll take a look. Thanks.
